### ISSUE TITLE
Fetchmore is not using cache

### ISSUE THREAD

**Issue by robertsmit** [apollographql/apollo-client#7430]
When using pagination and doing a fetchMore request the cache is not used to read data,

 While data is already in the cache and I implemented read and merge Field policies, it looks like the fetch more is not calling the read function and data is only showing up after an network request.
In the exterme example when I set fetchPolicy to 'cache-only', fetchmore is doing network requests. 

When you define and typePolicy whit a read function and doing a console.log on the connection field, you can see that it is not even called when doing fetchMore.

Apollo version 3.3

**Comment by benjamn** [comment]
@robertsmit Can you share the code for your field policy (including `read`, `merge`, and `keyArgs`) and `fetchMore` call?

**Comment by robertsmit** [comment]
> @robertsmit Can you share the code for your field policy (including `read`, `merge`, and `keyArgs`) and `fetchMore` call?
 I have added some code for clearance.

**Comment by benjamn** [comment]
@robertsmit If you have some way to detect the event of navigating back to the component, you should be able to call `setLimit(10)` to reset the window.

The `fetchMore` method creates a separate request that always has a fetch policy of `network-only`, which is why it doesn't try to read from the cache first.

**Comment by awlevin** [comment]
@benjamn I have a related question about `offsetLimitPagination`. We're implementing an infinite scrolling list where mutations can happen on items in the list (think: editing a comment in a news feed). Using `offsetLimitPagination` helps tremendously with reactively updating the UI and automating the feed concatenation, but it feels super hard to maintain. 

For example:
```js
typePolicies: {
  Query: {
    fields: {
      posts: offsetLimitPagination([/* this array has to contain every possible argument except for offset? otherwise it breaks "post" queries that aren't supposed to infinite scroll */])
    }
  }
}
```

To be a little more concrete: when I was forgetting a particular key, fetching the next page of the list would cause other pieces of UI to concatenate results where I wouldn't want them to. So if I want an infinite scrolling list, now I have to add `keyArgs` for every other combination of arguments for `posts` queries in our repo?

**Comment by robertsmit** [comment]
> @robertsmit If you have some way to detect the event of navigating back to the component, you should be able to call `setLimit(10)` to reset the window.
> 
> The `fetchMore` method sends a separate request that always has a fetch policy of `network-only`, which is why it doesn't try to read from the cache first.

With navigating away, I mean the component get unmount (in case of react). So when you mount the component again, the initial limit is 10, but there is no possible solution to use fetchmore to load more items from the cache.

**Comment by robertsmit** [comment]
@awlevin I had also this problem. My current solution is to generate some typePolicies based on the schema. New arguments to fields are generated in the type policies automatically. But offcourse, a negated keyArgs would be nice. Something like nonKeysArgs.

**Comment by benjamn** [comment]
@robertsmit I might call it `nonKeyArgs` for brevity, but I like the idea! Your other idea about allowing `fetchMore` to take a non-default `options.fetchPolicy` (rather than always using `no-cache`) is interesting too.

**Comment by benjamn** [comment]
Though I'm open to something like `nonKeyArgs`, I realized after writing my previous comment that you can currently provide a custom `keyArgs` function to implement whatever behavior you want:
```ts
new InMemoryCache({
  typePolicies: {
    SomeType: {
      fields: {
        someField: {
          keyArgs(args) {
            // return a string based on args
          },
        },
      },
    },
  },
})
```

**Comment by robertsmit** [comment]
@benjamn Yes I do this on some places but take care to make it deterministic. For example first order de args by name and then serialize.

**Comment by pedropmedina** [comment]
> The fetchMore method sends a separate request that always has a fetch policy of no-cache, which is why it doesn't try to read from the cache first.

@benjamn does this mean that implementing read/merge won't have effect on returning the data from cache first as fetchMore will still hit the network?

**Comment by m4riok** [comment]
Hello @benjamn,

Please confirm that fetchMore still  sends a separate request that always has a fetch policy of no-cache to date.  The documentation still illustrates that the use of fetchMore is the recommended way to go for implementing pagination, but if this is the case it sort of defeats the purpose of having a cache in the first place. 

I had to use setVariables on the ObservableQuery to achieve attempting to read from the cache first which is but a footnote in the documentation. If fetchMore stil behaves this way , other than using setVariables is there any other recommended way to trigger the read function in my typePolicies and the ObservableQuery to fire with new values ?

**Comment by hect1c** [comment]
Was there any update to this? 

I've only just realised this happens while having a custom merge, read function and maintaining the offset and limit in local state. It seems fetchMore always causes a network request when I already have data in the cache. 

@m4riok Can you provide details in the solution you mentioned or used to achieved the desired result?

**Comment by m4riok** [comment]
@hect1c  simply call setVariables in place of fetchMore and this will trigger your read function with the new variables. Check if the data is in the cache and if it is return it, otherwise return undefined which will trigger a network request to your graphql server. The new data will trigger your merge function which will merge it into the cache and the read will run again only this time the data will be in the cache and will be returned by your read implementation. This sort of implementation assumes that you are using cache-first as the strategy.  When your read returns anything other than undefined your subscription call back to the observable query will fire and your ui will update. 

In my implementation my merge ensures I have a sorted array of objects based on a date field, and my read attempts to read after or before a given cursor. This is specific to the kind of pagination you are implementing. Note that I am not using react so I'm not sure if sharing a specific snippet will help you but the principle outlined above applies to any framework specific implementation.

**Comment by iteratus** [comment]
Unfortunately still no update at least in the code:
https://github.com/apollographql/apollo-client/blob/fea2bab4e2c50ee96374ea27eb7b52358ccb59ed/src/core/ObservableQuery.ts#L431

Since I'm using React I'm not able to switch to `setVariable` as @m4riok suggests and like it's statet in the docs https://github.com/apollographql/apollo-client/blob/fea2bab4e2c50ee96374ea27eb7b52358ccb59ed/docs/source/pagination/offset-based.mdx#L153

I tried one build with this hardcoded value removed and it worked like a charm with `cache-first` policy (for my usecase at least and without regression testing of course).

It would really be nice if it's possible to introduce the default value `no-cache` which we can manually overwrite instead of simply hardcoding it. The `updateQuery` referenced in the comment (line 429) is optional, so it even would make sense to check whether or not `updateQuery` is set and only apply `no-cache` if it's really necessary, wouldn't it?

**Comment by dentuzhik** [comment]
@iteratus you mention in your [comment](https://github.com/apollographql/apollo-client/pull/11073#issuecomment-1713288177) of your closed PR that you have found some workaround using `useLazyQuery`. 

If it's not too much work, can you share what was that workaround?

**Comment by iteratus** [comment]
@dentuzhik Sure, let me try to explain it with words first.

We had the situation where we needed to load a list of 60 products in chunks of 12. By reaching the end of the first 12 products we would use `fetchMore` with an intersection observer to load the next 12. When reaching the end of 60 products, there was a load more button which starts the next 60 products in chunks of 12. So at first we passed `fetchMore` down to the component as a prop (if I remember correctly) which handles the request as well as to the button.

So to get to a solution, we ended up refactoring everything a bit. With `fetchMore` we already passed the variables `offset` and `amount` to the other components. But instead of calling `fecthMore` we now used `useLazyQuery` within them. The query itself was available through a regular import. As you know `useLazyQuery` does not execute immediately but returns a tuple with a query function. So the intersection observer simply calls this query function instead of `fetchMore`.

I'm still not super happy with the solution, but we needed to meet a deadline and it met our requirements. If it's still unclear I can create some code sandbox with both approches, just let me know.

**Comment by bigfree** [comment]
@iteratus, i have the same problem right now. I load a long list that combines an infinite scroll and button. I looking at network activity and fetch more still load from network even if the data is cached. I'll try to use your solution, but i don't like it much either.
